### PR TITLE
prov/shm: add iface parameter to smr_select_proto

### PR DESCRIPTION
**Pull request by wenduwan**
Pass through hmem iface to proto selection logic.

### CODE REVIEW

File: prov/shm/src/smr_ep.c [hunk 1]
```diff
@@ -634,13 +634,13 @@ static int smr_format_sar(struct smr_ep *ep, struct smr_cmd *cmd,
...
 
-int smr_select_proto(bool use_ipc, bool cma_avail,
-		     uint32_t op, uint64_t total_len, uint64_t op_flags)
+int smr_select_proto(enum fi_hmem_iface iface, bool use_ipc, bool cma_avail,
+                     uint32_t op, uint64_t total_len, uint64_t op_flags)
 {
 	if (op == ofi_op_read_req) {
 		if (use_ipc)
 			return smr_src_ipc;
-		if (cma_avail && FI_HMEM_SYSTEM)
```
Comment (wenduwan) [inline prov/shm/src/smr_ep.c:643]: I guess `cma_avail && FI_HMEM_SYSTEM` does not make much sense. Are we trying to test `cma_avail && FI_HMEM_SYSTEM == iface` instead?

Comment (aingerson) [inline prov/shm/src/smr_ep.c:643]: Woah, that's a big miss. Can you add something in the commit message about how this fixes a bug where CMA would never be selected for a read?

**Review by aingerson — APPROVED** [review]
Code change looks good. Would add something to the commit message about fixing/reenabling CMA for reads. Can you backport this as well please? Thank you for catching!

File: prov/shm/src/smr_ep.c (hunk 1 above)
Comment (aingerson) [reply prov/shm/src/smr_ep.c:643]: Yeah definitely a bug that this fixes. Thank you!

Comment (wenduwan) [reply prov/shm/src/smr_ep.c:643]: Thanks for confirming!

Comment (wenduwan) [reply prov/shm/src/smr_ep.c:643]: Thanks. Updated commit message. I will backport the change to 1.18.x branch.

Do we need it for 1.17.x as well?

Comment (aingerson) [reply prov/shm/src/smr_ep.c:643]: Well the only real use case of shm is efa so I'll let you decide how far back you want to go. The issue technically goes all the way back to 1.15.x

Comment (wenduwan) [reply prov/shm/src/smr_ep.c:643]: Cool then we only need to backport to 1.18.x

**Review by wzamazon — APPROVED** [review]

**Comment by wenduwan** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
Please back port

**Pull request merged**